### PR TITLE
Issue 1215/sign up and participant listing

### LINKED ISSUE #1215: 🤸‍♀️ Sign-up and participant listing
## Description
A user should be able to see a list of all the people who have signed up for an event and/or been booked. 

See epic #1214 for context.

## Relevant Job Stories
* When I’m using events to organize basic activism and one of the actvists who signed up would be suitable for running the event, I want to reach out to them, so that I can make sure they are able to do so

## Prerequisites
None

## Requirements
- [x] There should be two sections, Sign-ups and Booked participants (cancelled is handled by separate issue)
  - [x] The sign-up section should only be visible if there are pending sign-ups
- [x] Each section should have a title, a description, and a list/table
- [x] Each section header should have an icon
- [x] Each section header should have a number
  - [x] For sign-ups, the number should indicate the total number of sign-ups
  - [x] For booked participants, the numbers should indicate the current number of participants as well as the minimum required number as configured for the event
- [x] Each list should have six columns
  - [x] There should be an avatar column
  - [x] There should be a (single) name columns
  - [x] There should be a phone number column
  - [x] There should be an email column
  - [x] There should be a "notified" column, indicated when (if at all) the person has received a reminder
  - [x] There should be a column for action buttons
    - [x] Under "Sign-ups" the only action should be "book", which should add the person as a participant
    - [x] Under "Participants", the action should be empty for now (Cancelling tracked elsewhere)

## Possible implementations
Components:
* Use `DataGridPro` for the tables.
* Use `ZUIRelativeTime` for the "notified" column values.

API:
* Use `/orgs/ID/actions/ID/responses` for sign-ups
* Use `/orgs/ID/actions/ID/participants` for booked participants (including reminders)

## Design specifications
The mockup below contains much more than what this issue pertains to. Refer to "Requirements" above for the items needed to do for this issue, and to the mockup below only for visual guidance when completing the required items.

**BEWARE:** The mockup contains inconsistencies with regards to how many people are in a list vs what numbers are displayed in summaries etc. Ignore the information in the mockup and worry only about the style.

image

## Open questions
None

### PR DESCRIPTION
**Pull request by rebecarubio**
## Description
This PR adds the Sign-up and participant listing in Participants Tab.

## Screenshots
image

## Changes
* Adds sign-ups listing via DataGridPro component and the model function to get pending sign-ups
* Adds the booked Participant listing  via DataGridPro component and the model function to get participants
* Adds a new lint rule to check self-closed components (and fixes all files that didn't follow the rule)
* Changes participant and sign-ups calculations from ParticipantSummaryCard to the EventModel to make them reusable.
* Changes the participants page component to get the event data from the ZUIFuture
* Fixes a bug in participantAdded function in store and now the array gets updated after add a new participant

## Notes to reviewer
Cancel button gets fixed in a separate issue. Only booked button is working now.

## Related issues
Resolves #1215

### CODE REVIEW

File: src/features/events/store.ts
Comment (richardolsson) [inline src/features/events/store.ts:113]: Did it not work to just push? [The `immer` documentation](https://immerjs.github.io/immer/update-patterns/) seems to state that it should (note that the `produce()` part is irrelevant for us, it's handled by redux toolkit).

File: src/features/events/models/EventDataModel.ts
Comment (richardolsson) [inline src/features/events/models/EventDataModel.ts:49]: Why have these numbers as arguments? Couldn't the model just get the values using `this.getAvailParticipants()` and `this.getData().data.num_participants_required`?

File: src/features/events/models/EventDataModel.ts
Comment (richardolsson) [inline src/features/events/models/EventDataModel.ts:71]: Nice!

File: src/features/events/models/EventDataModel.ts
Comment (richardolsson) [inline src/features/events/models/EventDataModel.ts:38]: I think the naming of methods in this class are not entirely consistent. Some methods called `getXParticipants()` (or similar) return an array of participants, while others return a number. I would distinguish the methods returning numbers from the ones returning actual participants, using one of these patterns:

A. `getXParticipants()` returns list, `getNumXParticipants()` returns count, OR
B. `getXParticipants()` returns list, `getXParticipantCount()` returns count

File: src/features/events/models/EventDataModel.ts
Comment (richardolsson) [inline src/features/events/models/EventDataModel.ts:4]: It's not ideal to use the theme like this, or to use it inside a model at all. It will cause problems in case we make themes dynamic (which we will likely want to do to implement "night mode" at some point). But let's ignore that for now because we will likely abandon models before we implement night mode anyway.

File: src/features/events/components/EventParticipansTab/EventParticipantsList.tsx
Comment (richardolsson) [inline src/features/events/components/EventParticipansTab/EventParticipantsList.tsx:30]: This column could probably have a fixed width, I think?

File: src/features/events/components/EventParticipansTab/EventParticipantsList.tsx
Comment (richardolsson) [inline src/features/events/components/EventParticipansTab/EventParticipantsList.tsx:164]: It looks to me like this could be made a reusable component that is used twice (eventually three times) on this page. Does that make sense?

**Review by richardolsson — CHANGES_REQUESTED** [review]
Great work on this! I think it looks very good. 💯 

I only have some minor improvements that I think we should make, commented below! 👇

File: src/pages/organize/[orgId]/projects/[campId]/events/[eventId]/participants.tsx
Comment (richardolsson) [inline src/pages/organize/[orgId]/projects/[campId]/events/[eventId]/participants.tsx:60]: Oh I just noticed this little mistake as well. The empty tag is not closed as I would expect. Maybe it's a good idea to add the [`self-closing-comp` linting rule](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/self-closing-comp.md)? Would you mind adding it as part of this PR?

Comment (rebecarubio) [reply src/pages/organize/[orgId]/projects/[campId]/events/[eventId]/participants.tsx:60]: Great idea! I'll implement it!

File: src/features/events/store.ts
Comment (rebecarubio) [reply src/features/events/store.ts:112]: Yes! It's working! Nice to have less lines of code to do this :)

File: src/features/events/models/EventDataModel.ts
Comment (richardolsson) [reply src/features/events/models/EventDataModel.ts:38]: Did you do anything about this, @rebecarubio?

File: src/zui/ZUIEventParticipantLists.tsx
Comment (richardolsson) [inline src/zui/ZUIEventParticipantLists.tsx:58]: The name of this module and the name of the component do not match. The component is called `ZUIEventParticipantsList` (plural participants, singular list) but the module is called `ZUIEventParticipantLists` (singular participant, plural lists).

I also don't think this should be a ZUI component since I doubt that it would be used anywhere outside of the events feature. I would even just call it `ParticipantListSection` and put it in in the same module as where it's used, i.e. in `EventParticipantsList`.

File: src/features/events/components/EventParticipansTab/EventParticipantsList.tsx
Comment (richardolsson) [inline src/features/events/components/EventParticipansTab/EventParticipantsList.tsx:1]: The module (file) that this code is in is misspelled. I also don't think it should be inside a folder at all, just put it right at `src/features/events/components/EventParticipantsList.tsx`.

Comment (richardolsson) [reply src/features/events/components/EventParticipansTab/EventParticipantsList.tsx:1]: Please also remove the folder `EventParticipantsList` and put `AddPersonButton` directly in `components`. I realize this is not your code originally but I didn't realize when I reviewed #1267 that it was in this directory.

File: src/pages/organize/[orgId]/projects/[campId]/events/[eventId]/participants.tsx
Comment (richardolsson) [inline src/pages/organize/[orgId]/projects/[campId]/events/[eventId]/participants.tsx:52]: The `<AddPersonButton>` should be at the top of the screen.

## This PR now
image

## Design mockups
image

**Review by richardolsson — CHANGES_REQUESTED** [review]
We're very close now! I just found some code hygiene stuff, and a design issue after merging `main`. Comments below. 👇

File: src/features/events/components/EventParticipantsList.tsx
Comment (richardolsson) [inline src/features/events/components/EventParticipantsList.tsx:29]: Sorry for not realizing this sooner, but now that I see how everything's put together, it seems weird to me to define this out here. I feel like the columns should be defined inside of the component that renders the grid, i.e. in `ParticipantListSection`. How does that sound to you?

**Review by richardolsson — APPROVED** [review]
I think this looks good now. I would personally prefer the column definitions to be moved from `EventParticipantsList` to `ParticipantListSection`, but I guess that's a matter of personal taste so I'll let you decide if you want to do that.

**Review by richardolsson — APPROVED** [review]
Nice!! 🚀  This is ready to merge as soon as the tests pass. 🤞

**Review by richardolsson — APPROVED** [review]
Green light to merge! 🟢

**Pull request merged**
